### ISSUE TITLE
issue in bam2pairs

### ISSUE THREAD

**Issue by etiennejourdier** [koszullab/hicstuff#75]
Hi,
I have an issue in bam2pairs while executing the pipeline on dataset SRR5092176.
```
Traceback (most recent call last):
  File "/opt/conda/bin/hicstuff", line 8, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.9/site-packages/hicstuff/main.py", line 56, in main
    command.execute()
  File "/opt/conda/lib/python3.9/site-packages/hicstuff/commands.py", line 869, in execute
    hpi.full_pipeline(
  File "/opt/conda/lib/python3.9/site-packages/hicstuff/pipeline.py", line 855, in full_pipeline
    bam2pairs(bam1, bam2, pairs, info_contigs, min_qual=min_qual)
  File "/opt/conda/lib/python3.9/site-packages/hicstuff/pipeline.py", line 165, in bam2pairs
    end1.query_name = re.sub(end_regex, '', end1.query_name)
AttributeError: 'NoneType' object has no attribute 'query_name'
```
The command I used is
```
hicstuff pipeline -t 16 --no-cleanup -o out/ --prefix Rut1 -a bowtie2 --mapping iterative -g RutC30.fa 'SRR5092176_1.fastq.gz' 'SRR5092176_2.fastq.gz' --force
```
The error occurs with both mapping options, iterative or normal. The error also occurs while trying to restart the pipeline from the bam files

Because the valid.pairs file was complete, it seems to me that the issue occurs at the very end of the loop because the for and rev files do not have the same number of reads :
- 49,129,625 for SRR5092176_1.fastq.gz <=> Rut1.for.bam
- 49,129,632 for SRR5092176_2.fastq.gz <=> Rut1.rev.bam

I added a counter and checked that the issue indeed occurs only for the 7 latest reads which only exists in the rev file.
I managed to skip the error by replacing the command
```
 end1.query_name = re.sub(end_regex, '', end1.query_name)
```
by
```
try:
   end1.query_name = re.sub(end_regex, '', end1.query_name)
except AttributeError:
  print('error')
```
and the same for end2, but I don't know if it is the right way to handle this error ?

**Comment by cmdoret** [comment]
Hi @etiennejourdier, thanks for bringing this to our attention.

> the issue occurs at the very end of the loop because the for and rev files do not have the same number of reads 

Makes sense, but that should never happen. Do you know where this discrepancy comes from?

Unless there's a good reason why there may not be the same numbers, I think it makes sense to crash when this happens, as this could indicate a bigger problem.

**Comment by etiennejourdier** [comment]
Hi @cmdoret 

> Makes sense, but that should never happen. Do you know where this discrepancy comes from?
That's right, it should not happen. I [checked on SRA](https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&page_size=10&acc=SRR5092176&display=reads) and the problem was already there, The .1 reads are missing for the last reads. If you want to investigate further, you'll have to ask Martial, he was the one who deposited the file at the time. 😁
![image](https://github.com/koszullab/hicstuff/assets/115074079/fcb4a0bd-0912-4399-8561-34f2f8625f7c)

> it makes sense to crash when this happens, as this could indicate a bigger problem.
It's true, but as it shouldn't happen, it took me a long time to identify the problem. Maybe the script could check if the number of reads is consistent in the 2 files, and display a warning if it isn't?

**Comment by cmdoret** [comment]
@mmarbout https://github.com/koszullab/hicstuff/issues/75#issuecomment-1594246809 tagging you in case you want to take a look at that library

**Comment by etiennejourdier** [comment]
> One way this could be addressed is to issue a human-readable error message, what do you think?

It looks fine to me (I don't know how you'll detect that the error is due to an inconsistent number of reads, but I trust you!).

**Comment by js2264** [comment]
Fixed in a75f2c14752d36afea3242f27d54628dd29bf484 (will be part of [release 3.2.0](https://github.com/koszullab/hicstuff/milestone/1))
